### PR TITLE
[DataGrid] Fix date operators not working with date-time values

### LINKED ISSUE #1710: [DataGrid] Date operators not working with date-time values
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

If the value of a `date` or `dateTime` column has the hour and minutes set, most of the filter operators don't work. The only operators that return results are "is on or after/before", but they don't match with values that have the exact date ("is" part).

## Expected Behavior 🤔

Filter operators in `date` columns should consider only the date components (day, month and year). In `dateTime` columns, the operator should compare the date components and also the hour and minutes component, ignoring the seconds.
## Steps to Reproduce 🕹

Steps:

1. Open https://material-ui.com/components/data-grid/demo/#data-grid-demo
2. Apply a "is" filter on "Maturity Date" or "Trader Date"
3. Nothing is returned

6a8UUlrqae

## Context 🔦

Noticed it while working in #1708.

### PR DESCRIPTION
**Pull request by m4theushw**
Closes #1710 
To be merged after #1708

The fix for this bug involves resetting the time components of the date value before comparing it. The components we need to reset depend on the column type:
- If `date`, then resets the hours, minutes, seconds, and milliseconds
- If `dateTime`, then resets only seconds and milliseconds.

I also took the opportunity to remove a lot of duplicated code and move it to an abstraction.

### CODE REVIEW

File: packages/grid/_modules_/grid/models/colDef/gridDateOperators.ts [hunk 1]
```diff
@@ -2,24 +2,45 @@ import { GridFilterInputValue } from '../../components/panel/filterPanel/GridFil
...
+    }
+    // Make a copy of the date to not reset the hours in the original object
+    const valueAsDate = new Date(value instanceof Date ? value : value.toString());
+    const timeToCompare = valueAsDate.setHours(
+      showTime ? valueAsDate.getHours() : 0,
+      showTime ? valueAsDate.getMinutes() : 0,
+      0,
+      0,
+    );
+    return compareFn(timeToCompare, time);
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/colDef/gridDateOperators.ts:32]: Should it have?

```suggestion
    return compareFn(timeToCompare.getTime(), time);
```

**Review by oliviertassinari — COMMENTED** [review]
I would be careful with the performance. The ISO format is meant to allow string comparison to working with dates, maybe we could rely on it. Or maybe it could be more efficient to compare numbers with a modulo https://stackoverflow.com/questions/34722862/how-do-i-remove-time-part-from-javascript-date/34723015 or maybe we only have to care about removing the time for when the equality operator is involved 🤷‍♂️ 

The current version seems about 23% slower when filtering 100k rows.

File: packages/grid/_modules_/grid/models/colDef/gridDateOperators.ts (hunk 1 above)
Comment (m4theushw) [reply packages/grid/_modules_/grid/models/colDef/gridDateOperators.ts:39]: `timeToCompare` is already the number of milliseconds.

**Comment by m4theushw** [comment]
@oliviertassinari The `toISOString()` returns the date in UTC. We can't use it to compare with a local date. Another option would be to use `toLocaleDateString('en-US')` but I have to extract the values, convert to numbers and create the date object -> no performance gain. 

We could only not to care about the time when the operator is _before_ and the column type is `date`. I update this PR to check for this condition. But in general we need to reset the time, like this example with _after_:

**Scenario:**
**Given** the rows are 21/05/2021 04:00 and 22/05/2021 00:00
**When** I filter for rows AFTER 21/05/2021
**Then**  I expect to see only 22/05/2021 00:00

If I don't reset the time, I see 21/05/2021 04:00 and 22/05/2021 00:00, because 21/05/2021 04:00 is AFTER the filtered value.

> The current version seems about 23% slower when filtering 100k rows.

How did you obtain this value? I used `console.time()` and got very different times, sometimes faster than before the time reset, which doesn't make sense

**Comment by oliviertassinari** [comment]
> How did you obtain this value?

Open https://deploy-preview-1722--material-ui-x.netlify.app/components/data-grid/#commercial-version and measure the time it takes to filter the "Maturity Date" column.

File: packages/grid/_modules_/grid/models/colDef/gridDateOperators.ts (hunk 1 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/models/colDef/gridDateOperators.ts:39]: Oh wow, ok, we rely on the return value of setHours, smart.

**Comment by m4theushw** [comment]
@oliviertassinari I did a more systematic performance test measuring the time it took in `apiRef.current.applyFilters` to apply the filter in all rows. The difference seems negligible.

image

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 2]
```diff
@@ -222,10 +215,174 @@ describe('<DataGrid /> - Filter', () => {
...
+    ].forEach(({ operator, value, expected }) => {
+      it(`should work correctly with operator=${operator}`, () => {
+        render(
+          <TestCase
+            value={value}
+            operator={operator}
+            rows={[
+              {
+                id: 3,
+                brand: new Date(2000, 11, 1, 8, 30, 15, 20), // 12/1/2000, 8:30:15 AM
```
Comment (dtassone) [inline packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx:365]: Can we add a test with null, undefined, and empty values as well as dates?

Comment (m4theushw) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx:404]: I added more tests with undefined and empty values. We can't test for null or dates because the definition of the filter model doesn't allow them.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 3]
```diff
@@ -189,12 +189,51 @@ describe('<DataGrid /> - Filter', () => {
...
+      { operator: 'is', value: '' },
+      { operator: 'is', value: undefined },
+      { operator: '', value: '2000-12-01' },
+      { operator: '', value: '' },
+    ].forEach(({ operator, value }) => {
+      it(`should not filter when operator='${operator}' and value='${value}'`, () => {
+        render(
+          <TestCase
+            value={value}
+            operator={operator}
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx:201]: This was confusing when reviewing the PR, it would have been clearer with:

```suggestion
            operatorValue={operatorValue}
```

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 4]
```diff
@@ -189,12 +189,51 @@ describe('<DataGrid /> - Filter', () => {
...
-      { operator: 'is', value: new Date(2000, 11, 1), expected: ['12/1/2000'] },
-      { operator: 'not', value: new Date(2000, 11, 1), expected: ['1/1/2001', '1/1/2002'] },
-      { operator: 'after', value: new Date(2001, 0, 1), expected: ['1/1/2002'] },
-      { operator: 'onOrAfter', value: new Date(2001, 0, 1), expected: ['1/1/2001', '1/1/2002'] },
-      { operator: 'before', value: new Date(2001, 0, 1), expected: ['12/1/2000'] },
-      { operator: 'onOrBefore', value: new Date(2001, 0, 1), expected: ['12/1/2000', '1/1/2001'] },
+      { operator: 'is', value: '' },
+      { operator: 'is', value: undefined },
+      { operator: '', value: '2000-12-01' },
+      { operator: '', value: '' },
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx:195]: Is it possible for the `operatorValue` to be empty? In the UI, it doesn't seem possible.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx [hunk 5]
```diff
@@ -189,12 +189,51 @@ describe('<DataGrid /> - Filter', () => {
...
   describe('date operators', () => {
     [
-      { operator: 'is', value: new Date(2000, 11, 1), expected: ['12/1/2000'] },
-      { operator: 'not', value: new Date(2000, 11, 1), expected: ['1/1/2001', '1/1/2002'] },
-      { operator: 'after', value: new Date(2001, 0, 1), expected: ['1/1/2002'] },
-      { operator: 'onOrAfter', value: new Date(2001, 0, 1), expected: ['1/1/2001', '1/1/2002'] },
-      { operator: 'before', value: new Date(2001, 0, 1), expected: ['12/1/2000'] },
-      { operator: 'onOrBefore', value: new Date(2001, 0, 1), expected: ['12/1/2000', '1/1/2001'] },
+      { operator: 'is', value: '' },
+      { operator: 'is', value: undefined },
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx:193]: This is weird for the `value` to take in undefined. When empty, shouldn't it be `null` if the type is rich, like a multi-select and an empty string if the type is simple?

Comment (m4theushw) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx]: > This is weird for the value to take in undefined.

Using the `filterModel` prop it can be undefined.  

> When empty, shouldn't it be null if the type is rich, like a multi-select and an empty string if the type is simple?

Yes, and I would extend your suggestion saying that it should be typed as `any` and cast to the appropriated value: string for string columns, number for numeric columns and so on. We'll need to revisit this in #1320 to support an "one of" operator that allows to filter for rows that have one of the values specified.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx (hunk 4 above)
Comment (m4theushw) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx]: Using the `filterModel` prop it can be empty.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx (hunk 3 above)
Comment (m4theushw) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx]: Updated. IMHO, `operator` should be the name of the operator (is, after, before, contains) and `operatorValue` is the argument that feeds the operator.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx (hunk 5 above)
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx]: > Using the filterModel prop it can be undefined.

True, my point was that it doesn't seem to be something the types should allow.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx (hunk 2 above)
Comment (dtassone) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx:404]: I meant undefined value for a cell in a row, along with some dates, I believe you just put undefined on the filter value.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx (hunk 3 above)
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx]: Yeah, it's confusing, it took me a while to figure this out. Not changing the terminology in the test is already one step forward toward clarity :)

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx (hunk 2 above)
Comment (m4theushw) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx:474]: Updated with more test cases with null and empty inside the `rows` prop.

File: packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx (hunk 4 above)
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/filtering.DataGrid.test.tsx]: This sounds indeed like a good possible opportunity for follow-ups.

**Review by dtassone — APPROVED** [review]
